src/parser/core/modules/CooldownDowntime.tsx

```diff
@@ -0,0 +1,264 @@
+				// spend accumulated charges
+				count += charges
+				this.debugLog('Expected ' + charges + ' usages at ' + this.getDisplayTime(currentTime) + '. Count: ' + count)
+				charges = 0
+
+				// build a new charge at the next charge time
+				currentTime += step
+				charges += 1
+
+				// apply resets that are found
+				while (gResets.length > 0 && gResets[0].timestamp < currentTime) {
+					const rs = gResets[0]
+					const previousTime = currentTime
+					if (fullReset) {
+						if (charges < maxCharges) {
+							// if not at max charges, the full reset adds a charge and resets the cooldown timer
+							currentTime = rs.timestamp
+						} else {
+							// Used a reset while fully recharged - count the cooldown of the skill as lost reset time
+							timeLost += gRep.cooldown
+						}
+					} else if (currentTime - resetTime < rs.timestamp) {
+						if (charges < maxCharges) {
+							// if not at max charges, the "extra" reset time counts toward
+							// the next charge wihtout being lost
+							currentTime -= resetTime
+						} else {
+							timeLost += rs.timestamp - (currentTime - resetTime)
+							currentTime = rs.timestamp
+						}
+					} else {
+						currentTime -= resetTime
+					}
+					this.debugLog('Reset (' + rs.ability.name + ') used at ' + this.getDisplayTime(rs.timestamp) + '. Changing next charge time from ' + this.getDisplayTime(previousTime) + ' to ' + this.getDisplayTime(currentTime))
+					gResets.shift()
+				}
+
+				while (charges < maxCharges && this.downtime.isDowntime(currentTime)) {
+					this.debugLog('Saving charge during downtime at ' + this.getDisplayTime(currentTime) + '. ' + charges + ' charges  stored')
+
+					const window = this.downtime.getDowntimeWindows(currentTime)[0]
+					if (window.end < currentTime + step) {
+						count += charges
+						this.debugLog('Delayed charge spend at ' + this.getDisplayTime(window.end) + '. ' + charges + ' charges spent. No charge time lost. Count: ' + count)
+						charges = 0
+					}
+
+					currentTime += step
+					charges += 1
+				}
+
+				// full charges were built up during a downtime.  Move to the end of the downtime to spend charges.
+				if (this.downtime.isDowntime(currentTime)) {
+					const window = this.downtime.getDowntimeWindows(currentTime)[0]
+					this.debugLog('Downtime detected at ' + this.getDisplayTime(currentTime) + ' in window from ' + this.getDisplayTime(window.start) + ' to ' + this.getDisplayTime(window.end))
+					currentTime = window.end
+					// TODO: time after window end before usage.  should it just be first use offset? depends on what else was delayed and state in rotation
+				}
+			}
+			this.debugLog('Total count for group ' + gRep.name + ' is ' + count + '. Total reset time lost is ' + this.parser.formatDuration(timeLost) + '.')
+
+			return count
+		}
+
+	private getDisplayTime(currentTime: number): string {
+			return this.parser.formatDuration(currentTime - this.parser.fight.start_time)
```

Comment:
this is basically a dupe of `this.parser.formatTimestamp(currentTime)`

---

indentation here and below got funky